File: examples/elixir/user-server/test/user_types_test.exs
defmodule UserTypesTest do
  use ExUnit.Case, async: true

  describe "UserData struct" do
    test "can be created with all fields" do
      user = %User.UserData{
        user_id: "u-1",
        email: "alice@example.com",
        display_name: "Alice"
      }

      assert user.user_id == "u-1"
      assert user.email == "alice@example.com"
      assert user.display_name == "Alice"
    end

    test "display_name defaults to nil when only required fields are set" do
      user = %User.UserData{user_id: "u-1", email: "alice@example.com"}
      assert user.display_name == nil
    end

    test "supports struct update syntax" do
      user0 = %User.UserData{user_id: "u-1", email: "a@example.com", display_name: "A"}
      user1 = %{user0 | display_name: "Alice"}
      assert user1.user_id == "u-1"
      assert user1.display_name == "Alice"
    end
  end

  describe "GetUserInput struct" do
    test "requires user_id" do
      input = %User.GetUserInput{user_id: "u-1"}
      assert input.user_id == "u-1"
    end
  end
end

File: examples/erlang/user-server/test/user_types_test.erl
-module(user_types_test).

-include_lib("eunit/include/eunit.hrl").
-include("user_types.hrl").

%% -- user_data record ---------------------------------------------------------

user_data_all_fields_test() ->
    User = #user_data{
        user_id = <<"u-1">>,
        email = <<"alice@example.com">>,
        display_name = <<"Alice">>
    },
    ?assertEqual(<<"u-1">>, User#user_data.user_id),
    ?assertEqual(<<"alice@example.com">>, User#user_data.email),
    ?assertEqual(<<"Alice">>, User#user_data.display_name).

user_data_optional_display_name_test() ->
    User = #user_data{
        user_id = <<"u-1">>,
        email = <<"alice@example.com">>
    },
    ?assertEqual(undefined, User#user_data.display_name).

user_data_update_test() ->
    User0 = #user_data{
        user_id = <<"u-1">>,
        email = <<"a@example.com">>,
        display_name = <<"A">>
    },
    User1 = User0#user_data{display_name = <<"Alice">>},
    ?assertEqual(<<"u-1">>, User1#user_data.user_id),
    ?assertEqual(<<"Alice">>, User1#user_data.display_name).

%% -- get_user_input record ----------------------------------------------------

get_user_input_required_field_test() ->
    Input = #get_user_input{user_id = <<"u-1">>},
    ?assertEqual(<<"u-1">>, Input#get_user_input.user_id).
